Introduce database focus area, rename JavaScript focus area to JS & CSS, and phase out Site Health focus area

mukeshpanchal27 opened the pull request:
## Summary

PR that updates focus areas in the codebase
- Update perflab_get_focus_areas() to return the latest focus areas
- Rename javascript focus area to css-and-js (CSS & JS), including the directory in modules
- Move site-health modules into their new focus area directories
- Implement logic in perflab_get_module_settings() to migrate all old ("legacy") module slugs to the new ones (e.g. if you have site-health/audit-autoloaded-options in the DB, it should become database/audit-autoloaded-options)

<!-- Please reference the issue this PR fixes. If this PR does not fix the entire issue, change this to Addresses #... instead. -->
Fixes #554

## Relevant technical choices
To fix the unit test you have to update `default-enabled-modules.php` using https://make.wordpress.org/performance/handbook/performance-lab/releasing-the-plugin/#update-list-of-default-enabled-modules.

<!-- Please describe your changes. -->

## Checklist

- [ ] PR has either `[Focus]` or `Infrastructure` label.
- [ ] PR has a `[Type]` label.
- [ ] PR has a milestone or the `no milestone` label.

mukeshpanchal27 committed:
Update latest focus areas

mukeshpanchal27 committed:
Update codeowners

mukeshpanchal27 committed:
Remove Site Health from codeowners

mukeshpanchal27 committed:
Relocate existing modules

mukeshpanchal27 committed:
Update focus areas

mukeshpanchal27 committed:
Update unit tests

mukeshpanchal27 committed:
Update codeowners

mukeshpanchal27 committed:
Remove Site Health

mukeshpanchal27 committed:
Update legacy module code

mukeshpanchal27 committed:
Update default-enabled-modules.php

mukeshpanchal27 commented:
To fix the unit test related to `default-enabled-modules.php` I needs to update [7bdda84](https://github.com/WordPress/performance/pull/566/commits/7bdda84d268899e743dbd91868c2e5e0b76ac6a6) file manually. You can update it using https://make.wordpress.org/performance/handbook/performance-lab/releasing-the-plugin/#update-list-of-default-enabled-modules.

tillkruss commented on .github/CODEOWNERS:
```diff
@@ -60,6 +55,6 @@
-/modules/site-health/audit-full-page-cache                                  @manuelRod @westonruter
-/tests/modules/site-health/audit-full-page-cache                            @manuelRod @westonruter
+# Module: Audit Full Page Cache
+/modules/object-cache/audit-full-page-cache                                  @manuelRod @westonruter
+/tests/modules/object-cache/audit-full-page-cache                            @manuelRod @westonruter
```
This isn't really an `object-cache` module.

felixarntz replied:
@tillkruss I'd still say this is the closest focus area it fits into. Potentially we could formally broaden the focus to "Caching", if you're open to it. But even if not, I think for now this module fits best here, and to be fair, it will be removed soon anyway as it was already merged into core. :)

mukeshpanchal27 replied:
👍

tillkruss replied:
I don't think "Caching" makes sense. Let's go with this then.

felixarntz commented on .github/CODEOWNERS:
```diff
@@ -10,15 +10,10 @@
-/modules/site-health
-/tests/modules/site-health
-/tests/testdata/modules/site-health
+# Focus: CSS & JS
+/modules/css-and-js                         @aristath @sgomes
+/tests/modules/css-and-js                   @aristath @sgomes
+/tests/testdata/modules/css-and-js          @aristath @sgomes
```
Per the latest clarification, this should be "JS & CSS" (`js-and-css`) throughout.

mukeshpanchal27 replied:
Focus area updated throughout.

felixarntz commented on .github/CODEOWNERS:
```diff
@@ -35,22 +30,22 @@
-/modules/site-health/webp-support                                           @adamsilverstein @kirtangajjar @mitogh
-/tests/modules/site-health/webp-support                                     @adamsilverstein @kirtangajjar @mitogh
-/tests/testdata/modules/site-health/webp-support                            @adamsilverstein @kirtangajjar @mitogh
+# Module: WebP Support
```
Please revert the name, as the previous name `WebP Support Health Check` is the correct module name.

mukeshpanchal27 replied:
Module name reverted throughout.

felixarntz commented on .github/CODEOWNERS:
```diff
@@ -35,22 +30,22 @@
-/modules/site-health/audit-autoloaded-options                               @manuelRod
-/tests/modules/site-health/audit-autoloaded-options                         @manuelRod
-/tests/testdata/modules/site-health/audit-autoloaded-options                @manuelRod
+# Module: Audit Autoloaded Options
```
See above, please keep the previous name of the module. It should only be moved to its new focus area directory, not be renamed.

felixarntz commented on .github/CODEOWNERS:
```diff
@@ -35,22 +30,22 @@
-/modules/site-health/audit-enqueued-assets                                  @manuelRod
-/tests/modules/site-health/audit-enqueued-assets                            @manuelRod
-/tests/testdata/modules/site-health/audit-enqueued-assets                   @manuelRod
+# Module: Audit Enqueued Assets
```
See above regarding the module name.

felixarntz commented on .github/CODEOWNERS:
```diff
@@ -35,22 +30,22 @@
+# Module: Audit Enqueued Assets
+/modules/css-and-js/audit-enqueued-assets                                  @manuelRod
+/tests/modules/css-and-js/audit-enqueued-assets                            @manuelRod
+/tests/testdata/modules/css-and-js/audit-enqueued-assets                   @manuelRod
```
See above, these should be `js-and-css`.

felixarntz commented on .github/CODEOWNERS:
```diff
@@ -35,22 +30,22 @@
+/tests/testdata/modules/css-and-js/audit-enqueued-assets                   @manuelRod
 
-# Module: Persistent Object Cache Health Check
+# Module: Persistent Object Cache
```
See above regarding module name.

felixarntz commented on .github/CODEOWNERS:
```diff
@@ -60,6 +55,6 @@
-# Module: Full Page Cache Health Check
-/modules/site-health/audit-full-page-cache                                  @manuelRod @westonruter
-/tests/modules/site-health/audit-full-page-cache                            @manuelRod @westonruter
+# Module: Audit Full Page Cache
```
See above regarding module name.

felixarntz commented on admin/load.php:
```diff
@@ -184,11 +184,11 @@ function perflab_get_focus_areas() {
 		),
-		'javascript'   => array(
-			'name' => __( 'JavaScript', 'performance-lab' ),
+		'css-and-js'   => array(
+			'name' => __( 'CSS & JS', 'performance-lab' ),
```
See above, this should be `js-and-css`.

felixarntz commented on tests/admin/load-tests.php:
```diff
@@ -8,11 +8,11 @@
 
 	private static $demo_modules = array(
-		'javascript/demo-module-1' => array(
+		'css-and-js/demo-module-1' => array(
```
See above, please change this to `js-and-css` throughout.

felixarntz commented on load.php:
```diff
@@ -108,7 +108,23 @@ function perflab_get_module_settings() {
+
+	$legacy_module_slug = array(
+		'site-health/audit-autoloaded-options' => 'database/audit-autoloaded-options',
+		'site-health/audit-enqueued-assets'    => 'css-and-js/audit-enqueued-assets',
```
See above. Please also make sure to then rename the directories of the code for that module.

felixarntz commented on load.php:
```diff
@@ -108,7 +108,23 @@ function perflab_get_module_settings() {
 	// passed here, to support scenarios where this function is called before the 'init' action,
 	// for example when loading the active modules.
-	return (array) get_option( PERFLAB_MODULES_SETTING, perflab_get_modules_setting_default() );
+	$perflab_get_module_settings = (array) get_option( PERFLAB_MODULES_SETTING, perflab_get_modules_setting_default() );
```
Let's use a bit of a simpler variable name.
```suggestion
	$module_settings = (array) get_option( PERFLAB_MODULES_SETTING, perflab_get_modules_setting_default() );
```

felixarntz commented on load.php:
```diff
@@ -108,7 +108,23 @@ function perflab_get_module_settings() {
-	return (array) get_option( PERFLAB_MODULES_SETTING, perflab_get_modules_setting_default() );
+	$perflab_get_module_settings = (array) get_option( PERFLAB_MODULES_SETTING, perflab_get_modules_setting_default() );
+
+	$legacy_module_slug = array(
```
They are multiple slugs :)
```suggestion
	$legacy_module_slugs = array(
```

felixarntz commented on load.php:
```diff
@@ -108,7 +108,23 @@ function perflab_get_module_settings() {
+		'site-health/webp-support'             => 'images/webp-support',
+	);
+
+	foreach ( $perflab_get_module_settings as $module_slug => $module_data ) {
+		if ( isset( $legacy_module_slug[ $module_slug ] ) ) {
+			$perflab_get_module_settings[ $legacy_module_slug[ $module_slug ] ] = $module_data;
+			unset( $perflab_get_module_settings[ $module_slug ] );
+		}
+	}
+
+	return $perflab_get_module_settings;
```
Instead of looping through all the module settings, let's loop through just the legacy module slugs, as that is more efficient. It also makes the code easier to understand.
```suggestion
	foreach ( $legacy_module_slugs as $legacy_slug => $current_slug ) {
		if ( isset( $module_settings[ $legacy_slug ] ) ) {
			$module_settings[ $current_slug ] = $module_settings[ $legacy_slug ];
			unset( $module_settings[ $legacy_slug ] );
		}
	}

	return $module_settings;
```

mukeshpanchal27 replied:
Thanks for the review. This change make more sense.

felixarntz commented on load.php:
```diff
@@ -108,7 +108,23 @@ function perflab_get_module_settings() {
+	}
+
+	return $perflab_get_module_settings;
 }
```
Can you please add test coverage for this change? We should add at least one test that sets the legacy module slug in the DB, and then verify that this function still returns the same data for the new module slug.

We could use a `dataProvider` for that test in order to cover all 4 migrations.

mukeshpanchal27 replied:
Thank you for the feedback Felix. Unit test added https://github.com/WordPress/performance/pull/566/commits/ce593f4c9a629cd59cf4ad5d053d819569f44b5e

felixarntz reviewed (commented):
@mukeshpanchal27 This looks solid overall, I mostly have minor feedback on updating the `css-and-js` focus area to `js-and-css`, and a few things to improve. The most important thing missing would be to add test coverage for the module settings migration logic.

mukeshpanchal27 committed:
Merge branch 'trunk' into fix/554-update-focus-areas-codebase

mukeshpanchal27 committed:
Update focus area to js-and-css

mukeshpanchal27 committed:
Revert module name changes

mukeshpanchal27 committed:
Addressed code review

mukeshpanchal27 committed:
Add unit test

mukeshpanchal27 commented on tests/load-tests.php:
```diff
@@ -97,6 +97,32 @@ function( $default, $option, $passed_default ) use ( &$has_passed_default ) {
+		update_option( PERFLAB_MODULES_SETTING, $new_value );
+
+		$settings = perflab_get_module_settings();
+		$this->assertArrayNotHasKey( $legacy_module_slug, $settings, 'The settings do not contain the old legacy module slug in the database' );
+		$this->assertArrayHasKey( $current_module_slug, $settings, 'The settings contain an updated module slug in the database' );
```
As good habit like WP core, 

Test methods with multiple assertions need a `message` parameter for each assertion, per [Core Handbook - Writing PHPUnit Tests - Using Assertions](https://make.wordpress.org/core/handbook/testing/writing-phpunit-tests/#using-assertions).

mehulkaklotar commented on tests/load-tests.php:
```diff
@@ -97,6 +97,32 @@ function( $default, $option, $passed_default ) use ( &$has_passed_default ) {
 	}
 
+	/**
+	 * @dataProvider data_legacy_modules
```
@mukeshpanchal27 Missing `@test` here.

mukeshpanchal27 replied:
@mehulkaklotar To run test we can use two methods.
- At starting of function name use `test_`
- Or you can use the `@test` annotation in a method’s docblock.

mehulkaklotar commented on tests/load-tests.php:
```diff
@@ -97,6 +97,32 @@ function( $default, $option, $passed_default ) use ( &$has_passed_default ) {
+		$this->assertArrayHasKey( $current_module_slug, $settings, 'The settings contain an updated module slug in the database' );
+	}
+
+	public function data_legacy_modules() {
```
@mukeshpanchal27 can you add function description here like `Data provider for legacy modules` and what the function returns via `@return array`?

mukeshpanchal27 replied:
Thanks for the feedback! Docblock added for the `data_legacy_modules()`

felixarntz commented on .github/CODEOWNERS:
```diff
@@ -61,5 +56,5 @@
-/modules/site-health/audit-full-page-cache                                  @manuelRod @westonruter
-/tests/modules/site-health/audit-full-page-cache                            @manuelRod @westonruter
+/modules/object-cache/audit-full-page-cache                                  @manuelRod @westonruter
+/tests/modules/object-cache/audit-full-page-cache                            @manuelRod @westonruter
```
Nit-pick, but please keep the indentation of the owner usernames here aligned between the different modules, similar to how it was before (throughout the entire file).

mukeshpanchal27 replied:
Thanks for the review. indentation corrected.

felixarntz commented on tests/load-tests.php:
```diff
@@ -97,6 +97,32 @@ function( $default, $option, $passed_default ) use ( &$has_passed_default ) {
+	 * @dataProvider data_legacy_modules
+	 */
+	public function test_legacy_module_for_perflab_get_module_settings( $legacy_module_slug, $current_module_slug ) {
+		$new_value = array(
+			'site-health/audit-autoloaded-options' => array( 'enabled' => true ),
+			'site-health/audit-enqueued-assets'    => array( 'enabled' => true ),
+			'site-health/audit-full-page-cache'    => array( 'enabled' => true ),
+			'site-health/webp-support'             => array( 'enabled' => true ),
+		);
```
You can simplify this, you don't need to re-include the list of legacy module slugs here again. Only working with the single module that is being tested is sufficient.
```suggestion
		$new_value = array( $legacy_module_slug => array( 'enabled' => true ) );
```

felixarntz reviewed (approved):
@mukeshpanchal27 Looks great! I'll LGTM this already, but please address the two minor remaining quirks (there strictly not a blocker though).

mukeshpanchal27 committed:
Add docblock for data_legacy_modules()

mukeshpanchal27 committed:
Correct indentation

mukeshpanchal27 committed:
Update tests/load-tests.php

Co-authored-by: Felix Arntz <felixarntz@users.noreply.github.com>

mukeshpanchal27 commented:
@felixarntz I have addressed the feedback. Github shows the unit test related error for `Fatal error: Uncaught Error: Class 'WP_Image_Editor_Imagick' not found in /var/www/html/wp-content/plugins/performance/tests/testdata/modules/images/webp-uploads/class-wp-image-doesnt-support-webp.php:14` https://github.com/WordPress/performance/actions/runs/3374928930/jobs/5601071400 I think it was introduce because of latest WP update. I think all new PR changes fails now.

@adamsilverstein Do we needs to open new issue?

felixarntz commented:
@mukeshpanchal27 Thanks! Looks like there's already #572, let's prioritize getting that through so that then we can refresh this PR here to make tests pass again.

mehulkaklotar reviewed (approved):
@mukeshpanchal27 LGTM! 🚀

mukeshpanchal27 committed:
Merge branch 'trunk' into fix/554-update-focus-areas-codebase
